session.go

```diff
@@ -0,0 +1,172 @@
+	return nil
+}
+
+func sessionSettingValue(k SessionSetting, resp *Response) (interface{}, error) {
+	if resp == nil {
+		return nil, fmt.Errorf("unexpected session settings response: got nil")
+	}
+
+	if len(resp.Data) == 0 {
+		return nil, fmt.Errorf("session setting %s not found", k)
+	}
+
+	if len(resp.Data) != 1 {
+		return nil, fmt.Errorf("unexpected session settings response length %d",
+			len(resp.Data))
+	}
+
+	tuple, ok := resp.Data[0].([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("unexpected session settings response format: expected tuple, got %v",
+			resp.Data)
+	}
+
+	// (key, value) tuple expected
+	if len(tuple) < 2 {
+		return nil, fmt.Errorf("unexpected session settings response format: too few fields in the tuple (got %v)",
+			resp.Data)
+	}
+
+	return tuple[sessionSettingValueField], nil
+}
+
+func sessionSettingsValue(resp *Response) (map[SessionSetting]interface{}, error) {
+	if resp == nil {
+		return nil, fmt.Errorf("unexpected session settings response: got nil")
+	}
+
+	if len(resp.Data) == 0 {
+		return nil, fmt.Errorf("unexpected session settings response: got no rows")
+	}
+
+	val := make(map[SessionSetting]interface{}, len(resp.Data))
+
+	for _, item := range resp.Data {
+		tuple, tok := item.([]interface{})
+		if !tok {
+			return nil, fmt.Errorf("unexpected session settings response format: expected tuple, got %v", item)
+		}
+
+		// (key, value) tuple expected
+		if len(tuple) < 2 {
+			return nil, fmt.Errorf("unexpected session settings response format: too few fields in the tuple (got %v)", tuple)
+		}
+
+		skey, ok := tuple[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("unexpected session settings response key: %v", tuple[0])
+		}
+
+		val[SessionSetting(skey)] = tuple[1]
+	}
+
+	return val, nil
+}
+
+func wrapSessionRequestError(err error) error {
+	if errors.Is(err, ClientError{Code: ErrSpaceNotFound}) {
+		err = fmt.Errorf("session settings are not supported: %w", err)
+	}
+
+	return err
+}
+
+func (conn *Connection) SetSessionSetting(k SessionSetting, v interface{}) (interface{}, error) {
+	req := NewUpdateRequest(sessionSettingsSpace).
+		Key(sessionSettingKey{k}).
+		Operations(NewOperations().Assign(sessionSettingValueField, v))
+
+	resp, err := conn.Do(req).Get()
+	if err != nil {
+		return nil, wrapSessionRequestError(err)
+	}
+
+	return sessionSettingValue(k, resp)
+}
+
+func (conn *Connection) SessionSetting(k SessionSetting) (interface{}, error) {
+	req := NewSelectRequest(sessionSettingsSpace).
+		Key(sessionSettingKey{k}).
+		Limit(1)
+
+	resp, err := conn.Do(req).Get()
+	if err != nil {
+		return nil, wrapSessionRequestError(err)
+	}
+
+	return sessionSettingValue(k, resp)
+}
+
+func (conn *Connection) SessionSettings() (map[SessionSetting]interface{}, error) {
+	req := NewSelectRequest(sessionSettingsSpace).
+		Limit(1000)
+
+	resp, err := conn.Do(req).Get()
+	if err != nil {
+		return nil, wrapSessionRequestError(err)
+	}
+
+	return sessionSettingsValue(resp)
+}
```

Comment:
`NewErrorMarshalingRequest` is definitely a counterintuitive name. I'll give a thought about it in Monday, but for I think that request approach is much confusing: in current one user may not even know that "set session settings" is to send a request.

Reply:
> do we need a separate request for each option
No, one should be enough. Or two, if we decide to separate setters and getters. We also need to think how to carefully process `select{nil}` = "give all settings" case which also can be useful.

Reply:
> request approach is much confusing

On the one hand, a separate method is simpler to handle, on the other hand, it is more difficult to extend. For example, is there a way to use [context ](https://pkg.go.dev/context) with new methods or handle a [push message](https://www.tarantool.io/ru/doc/latest/reference/reference_lua/box_session/push/)?

That is why It is better to do everything in separate requests, if possible. Even if specialized methods seem more useful.

Reply:
> specialized requests will help to replace interface{} in setters by meaningful types

Since `_session_settings` format is `string, any`, I think using interface here is ok.

Reply:
But in practice, each setting has its own value type.

Reply:
One more thing to discuss here is a response. Current version may provide some user-friendly API (call get setting -- receive setting value), request one will have tuples in response (which may be better in case someone decide to extend session settings space with additional field).

Reply:
> request one will have tuples in response

I don't see this as a problem, it's true for all responses right now.

Reply:
See new API in current PR version. My thoughts:
- using `Setting.Set()`/`Setting.Get()` sounds like a bad approach at least because we already have `Get()` in a future and it may be confusing;
- using a single request for set/get is also confusing.

Reply:
I also think that introducing new request types are not worth it: less reusability, more copypaste, and no obvious profit.

Reply:
> less reusability

I don't agree.

> and no obvious profit.

Now request types are used to extend functionality:

1. https://github.com/tarantool/go-tarantool/issues/48
2. https://github.com/tarantool/go-tarantool/issues/67
3. We can add something else in the future.

In `v2`, I plan to add methods for decoding the header and body to requests so that there is no need to modify the code for a new requests:

https://github.com/tarantool/go-tarantool/issues/237#issuecomment-1346703612

Therefore it is better to make separate types.

> more copypaste

This is a temporary problem. Most copy-pastewill be deleted after we support go 1.18+ (generics).

Reply:
> Most copy-paste will be deleted after we support go 1.18+ (generics).

I don't see how it will work here.

Reply:
> We can extend the request functionality independently.

And if something will change in select/update, we will need to update copypasted code in several places.

Reply:
> There will be three different ones in different modules:
> 
> ```go
>     connector.Do(NewBlaBlaRequest(arg1).Arg(arg2))
>     connector.Do(SomeMethodName(arg1).Arg(arg2))
>     connector.Do(SomeRequest{arg1, arg2})
> ```

If you're pointing out the "New" and "Request" naming pattern, I did it so it won't be named "NewSetErrorMarshalingEnabledRequest", but if you insist, I may do so.

Reply:
I don't get the third one, though -- my implementations are functions.

Reply:
> > I don't agree. We can extend the request functionality independently.
> 
> It's essentially a Select (Update) request. To make it work, I need to copypaste all Select (Update) related code.

You are over-dramatizing.

```Go
type SettingsRequest struct {
    select *SelectRequest
    update *UpdateRequest
}

// new functions + methods for tarantool.Request interface + Context()

NewSetErrorMarshalingEnabledRequest(value bool) *SettingRequest {
    return newSetSettingsRequest(ErrorMarshalingEnabled, value)
}

// ...
```

Yes, you will have to copy-paste methods for tarantool.Request interface, but we hide the implementation from the user and we can always change it in the future. As example, it is very easy to replace EvalRequest by CallRequest instade the BroadcastRequest.

https://github.com/tarantool/go-tarantool/blob/master/watch.go#L7-L53

In addition, we hide unnecessary methods of SelectRequest and UpdateRequest from the user.

Reply:
Reworked based on your proposition